### DISCUSSION TITLE
How can I import a native C++ extension

### DISCUSSION THREAD

**Discussion by ngfelixl** [eclipse-theia/theia#14220] [Q&A] [answered]
Hi there. This is my first contribution to this repository, I hope it's in the correct place.

I am trying to integrate a NodeJS C++ Extension. The extension is a dedicated npm package, it provides an __index.js__ and index.d.ts file and this __index.js__ file _requires_ a native C++ extension. The repository structure is as follows (I am skipping irrelevant files):

```
build/Release/myaddon.node (and other files)
dist/index.js
src/
  myaddon.cpp
  index.ts
package.json
some.dll
test.ts
```

So the generated __dist/index.js__ file contains `require('../build/Release/myaddon.node')`. The __package.json__ points to this __index.js__ file. If I import it in the __test.ts__ file, I am able to execute this C++ code without any problems. The C++ extension loads a __.dll__ file located in the package root.

Now I cloned the Theia-Blueprint repository and executed the `yo theia-extension --standalone` "Backend Communication" and moved it to the `theia-extensions` folder. Without the C++ extension this extension integrates properly. When I add the C++ extension as a dependency to the extension and run `yarn electron build` it throws an error.

```
ERROR in ../../../nodesjs-libraries/myaddon/build/Release/myaddon.node 1:2
Module parse failed: Unexpected character '�' (1:2)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
(Source code omitted for this binary file)
 @ ../../../nodesjs-libraries/myaddon/dist/index.js 4:0-56
 @ ./node_modules/theia-ide-myextension-ext/lib/browser/myextension-contribution.js 28:31-72
 @ ./node_modules/theia-ide-myextension-ext/lib/browser/myextension-frontend-module.js 13:38-78
 @ ./src-gen/frontend/index.js 144:30-111

webpack 5.94.0 compiled with 1 error in 23145 ms
```

So it seems to find the file, but it's not _loaded_ correctly. The package is used in the __hello-backend-service.ts__ as follows:

```ts
import { injectable } from '@theia/core/shared/inversify';
import { HelloBackendService } from '../common/protocol';
import { addon } from 'myaddon-package-name';

@injectable()
export class HelloBackendServiceImpl implements HelloBackendService {
    sayHelloTo(name: string): Promise<string> {
        return new Promise<string>(resolve => {
            addon.add(5, 4, (_: unknown, data: unknown) => {
                resolve(`${name} ${data}`);
            });
        });
    }
}
```

I've installed it via `yarn link` and `yarn link "myaddon-package-name"` (with @ scope in reality). There is a symlink in both node_modules (root and the theia-extension one). It seems to be related to the Webpack bundling. I've tried to add the addon to the NativeWebpackPlugin which Theia provides in __applications/electron/webpack.config.js__, l.32:

```js
if (nodeConfig.config.optimization) {
    // Some TerserPlugins
}
nodeConfig.nativePlugin.nativeBinding('myaddon', 'myaddon-package-name/build/Release/myaddon.node');
```

And there is also a "node-loader" in place somehow and I verified the existence of __myaddon.node__. Unfortunately, there is no positive outcome, yet. Do you have any ideas? Thanks in advance!

Resources:
- https://github.com/eclipse-theia/theia/blob/master/dev-packages/native-webpack-plugin/src/native-webpack-plugin.ts
- https://github.com/theia-ide/theia-apps/tree/master/theia-cpp-electron (incomplete)

**Comment by msujew** [comment, marked as answer]
Hey @ngfelixl,

Thanks for the logs/context on this! Looking at this part:

```
@ ./node_modules/theia-ide-myextension-ext/lib/browser/myextension-contribution.js 28:31-72
@ ./node_modules/theia-ide-myextension-ext/lib/browser/myextension-frontend-module.js 13:38-78
```

It looks like you're attempting to load your native node package in the browser bundle. This is also indicated by the webpack error (we only generate a loader for `.node` files for the backend).

You need to bind the backend contribution in a dedicated `backend-module.ts` file that is entered as a `backend` module in your `package.json`. See for example:

https://github.com/eclipse-theia/theia/blob/1a7486e492145c3d6f9aafbeca3b7e329eef8fe1/packages/core/package.json#L145

**Reply by ngfelixl** [reply]
Awesome, thanks for your help. I was using outdated build artefacts, because I modified the dependency path with "file:" in the __package.json__ to avoid the failing installation (which happens when using the theia-blueprint). The extension itself is working now and executes C++ code properly. However the library is loading a __.dll__ file with the following code. It is fully working in the packages __test.ts__ file.

```cpp
#include <windows.h>
#include <string>
#include "napi.h"

HINSTANCE lib = LoadLibrary(TEXT("my.dll"));

if (lib != NULL) {
  // GetProcAddress calls.
}
```

My assumption is, that the path to the __my.dll__ file is broken. The __.dll__ file is bundled with the npm package in the packages root directory. But it seems that theia creates copies of the __.node__ files in the __examples/electron/lib/backend/native__ directory. Copying the __.dll__ won't work. Maybe I have to link the __.dll__ file in the __bindings.gyp__ of the library to be _included_ in the __.node__ file?

Maybe you or someone can answer me this off-topic question as well:
- Probably I should use the eclipse-theia/theia repo for development of new extensions, correct? Should I use the eclipse-theia/theia-blueprint when building the electron app, or can I create the electron .exe file also from the eclipse-theia/theia repo?

**Reply by msujew** [reply]
The generated webpack config for the backend shouls generally take care of copying the `.node` file. What kind of error are you encountering?

> Probably I should use the eclipse-theia/theia repo for development of new extensions, correct? Should I use the eclipse-theia/theia-blueprint when building the electron app, or can I create the electron .exe file also from the eclipse-theia/theia repo?

To develop your own app based on Theia, please take a look at [our documentation](https://theia-ide.org/docs/composing_applications/).

**Reply by ngfelixl** [reply]
As a workaround for the `LoadLibrary` problem: It works when using the absolute path. I will submit a solution, when I have one for relative paths.
